### PR TITLE
Implement PSATD solver in C++

### PR DESCRIPTION
**Pull request by RemiLehe**
## Overview

This implements the PSATD solver in C++, as an alternative to the existing Fortran code in PICSAR. (This is part of a more general trend to progressively move away from Fortran.)

More specifically, this PR adds a new user input parameter `psatd.hybrid_mpi_decomposition`. When set to 1, WarpX will use the PICSAR code with hybrid FFT. When set to 0 (default), WarpX uses the new C++ code, which only does local FFT in each box (for the moment).

## Limitations

This implementation is a first step, and has the following limitations **which will be addressed in later pull requests**:

- When running on GPU, the code should call cuFFT. (The current code is not functional on GPU.) - see issue #96 
- The code currently uses complex-to-complex FFT. Because the fields in real space are real, the performance would benefit from using the real-to-complex FFT instead. - see PR #98 
- When running on CPU, the current code does not use multi-threading for the FFT.
- The current code only works with double precision (see e.g. the definition of the `Complex` type).
- Despite the name of the parameter `hybrid_mpi_decomposition`, the current code still uses the dual decomposition (i.e. the field `Efield_fp_fft` instead of `Efield_fp`). A later PR will modify the code so as to use the fields `Efield_fp` directly. - see PR #99 
- The current code will probably not work with load-balancing and/or mesh refinement.
- The current code only performs local FFT. Eventually, we would like to move the hybrid-FFT code to C++ as well.

## Overall structure

When reviewing this PR, I would recommend to look at it in the following order:

**High-level:** The code in `WarpXFFT.cpp` uses the object `SpectralSolver` which performs the Fourier transforms and advances the fields.

**Lower level:** All the code for `SpectralSolver` is contained in the new folder `SpectralSolver`. This defines several lower-level classes which are members of the `SpectralSolver` class (see `SpectralSolver.H`)

### CODE REVIEW

**Comment by RemiLehe** [comment]
Note regarding the latest commit: I had made a mistake in the sign of the shift (when the FFT is performed from a cell-centered grid). 

It is surprisingly hard to identify this mistake in the 3D periodic Langmuir test **when looking only at the E field** (the E field looks valid with either signs). However, the error is clearly seen when looking at the B field:

- Previous sign of the shift:
B_field_shifted
By looks asymmetric, and is much too high

- Corrected sign of the shift:
B_field
By looks symmetric.

File: Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+/* \brief Transform the component `i_comp` of MultiFab `mf`
+ *  to spectral space, and store the corresponding result internally
+ *  (in the spectral field specified by `field_index`) */
+void
+SpectralFieldData::ForwardTransform( const MultiFab& mf,
+                                     const int field_index,
+                                     const int i_comp )
+{
```
Comment (MaxThevenet) [inline Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp]: May be worth it to add `BL_PROFILE("SpectralFieldData::ForwardTransform()");` here and in `BackwardTransform`?

File: Source/FieldSolver/SpectralSolver/SpectralKSpace.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        // Allocate shift coefficients
+        shift.resize( k.size() );
+
+        // Fill the shift coefficients
+        Real sign = 0;
+        switch (shift_type){
+            case ShiftType::TransformFromCellCentered: sign = -1.; break;
+            case ShiftType::TransformToCellCentered: sign = 1.;
+        }
+        constexpr Complex I{0,1};
```
Comment (MaxThevenet) [inline Source/FieldSolver/SpectralSolver/SpectralKSpace.cpp]: We might define it in `WarpX_complex.H`, but that's really a tiny detail

File: Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+#else
+        // Create FFTW plans
+        forward_plan[mfi] =
+            // Swap dimensions: AMReX FAB are Fortran-order but FFTW is C-order
+#if (AMREX_SPACEDIM == 3)
+            fftw_plan_dft_3d( bx.length(2), bx.length(1), bx.length(0),
+#else
+            fftw_plan_dft_2d( bx.length(1), bx.length(0),
+#endif
+            reinterpret_cast<fftw_complex*>( tmpRealField[mfi].dataPtr() ),
```
[the remark this answers is not in the document]
Comment (RemiLehe) [reply Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp]: Yeah, @WeiqunZhang suggested to use this. I did not know about this either.

File: Source/FieldSolver/SpectralSolver/PsatdAlgorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                X3(i,j,k) = - c*c * dt*dt / (3.*ep0);
+            }
+        });
+    }
+};
+
+/* Advance the E and B field in spectral space (stored in `f`)
+ * over one time step */
+void
+PsatdAlgorithm::pushSpectralFields(SpectralFieldData& f) const{
```
[the remark this answers is not in the document]
Comment (RemiLehe) [reply Source/FieldSolver/SpectralSolver/PsatdAlgorithm.cpp]: Hm, good point, but maybe this should be added in `WarpXFFT.cpp`, around the calls for the Fourier transform and the calls for the field update.
(That way, if we later implement other `Algorithm` objects, we do not need to add `BL_PROFILE` in each of them.)
What do you think?

File: Source/FieldSolver/SpectralSolver/SpectralKSpace.H [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+/* \brief Class that represents the spectral space.
+ *
+ * (Contains info about the size of the spectral space corresponding
+ * to each box in `realspace_ba`, as well as the value of the
+ * corresponding k coordinates)
+ */
+class SpectralKSpace
+{
+    public:
+        amrex::BoxArray spectralspace_ba;
```
[the remark this answers is not in the document]
Comment (RemiLehe) [reply Source/FieldSolver/SpectralSolver/SpectralKSpace.H]: `spectral_ba` is a new BoxArray that is created and defined in the constructor of `SpectralKSpace` (i.e. different from the `realspace_ba` which is passed to the constructor), 
so it needs to be stored somewhere (e.g. as a member of the class, here.)

On the other hand, `dm` is unchanged. We could store a **reference** to `dm` as a member of `SpectralKSpace`, but that seemed unnecessary to me, and maybe potentially confusing.

File: Source/FieldSolver/SpectralSolver/SpectralSolver.H [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+/* \brief Top-level class for the electromagnetic spectral solver
+ *
+ * Stores the field in spectral space, and has member functions
+ * to Fourier-transform the fields between real space and spectral space
+ * and to update fields in spectral space over one time step.
+ */
+class SpectralSolver
+{
+    private:
```
[the remark this answers is not in the document]
Comment (RemiLehe) [reply Source/FieldSolver/SpectralSolver/SpectralSolver.H]: Yes, I agree. I'll update the PR.

File: Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp (hunk 1 above)
Comment (RemiLehe) [reply Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp]: Yes, definitely! Any preference as to whether to add this around the call in `SpectralSolver`, or here?

File: Source/FieldSolver/SpectralSolver/WarpX_Complex.H [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#ifndef WARPX_COMPLEX_H_
```
[the remark this answers is not in the document]
Comment (RemiLehe) [reply Source/FieldSolver/SpectralSolver/WarpX_Complex.H]: Yes, I that makes sense. I'll update the PR accordingly.

**Comment by RemiLehe** [comment]
Thanks for the comments and feedback @MaxThevenet.
Regarding the examples, there is no specific example, but the Langmuir standard input can be run with WarpX compiled with `USE_PSATD=TRUE` for instance. (See e.g. https://github.com/ECP-WarpX/WarpX/blob/master/Regression/WarpX-tests.ini#L213)

File: Source/FieldSolver/SpectralSolver/SpectralKSpace.H (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/FieldSolver/SpectralSolver/SpectralKSpace.H]: Ok, that makes sense

File: Source/FieldSolver/SpectralSolver/PsatdAlgorithm.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaxThevenet) [reply Source/FieldSolver/SpectralSolver/PsatdAlgorithm.cpp]: Both are fine, I tend to think that `BL_PROFILE` is cheap and put it in all routines that may take time to easily detect issues, but I am not sure it is required indeed.

File: Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp (hunk 1 above)
Comment (MaxThevenet) [reply Source/FieldSolver/SpectralSolver/SpectralFieldData.cpp]: Right, probably better to put it in `SpectralSolver` for future developments

**Pull request merged**